cardinal/ecs/ecb/ecb.go

```diff
@@ -0,0 +1,501 @@
+		return err
+	}
+
+	toArchID, err := m.getArchIDForComponentsOrMakeIt(toComps)
+	if err != nil {
+		return err
+	}
+	fromArchID, err := m.getArchIDForComponentsOrMakeIt(fromComps)
+	if err != nil {
+		return err
+	}
+	return m.moveEntityByArchetype(fromArchID, toArchID, id)
+}
+
+// RemoveComponentFromEntity removes the given component from the given entity. An error is returned if the entity
+// does not have the component.
+func (m *Manager) RemoveComponentFromEntity(cType component.IComponentType, id entity.ID) error {
+	comps, err := m.GetComponentTypesForEntity(id)
+	if err != nil {
+		return err
+	}
+	var newCompSet []component.IComponentType
+	found := false
+	for _, comp := range comps {
+		if comp.ID() == cType.ID() {
+			found = true
+			continue
+		}
+		newCompSet = append(newCompSet, comp)
+	}
+	if !found {
+		return storage.ErrorComponentNotOnEntity
+	}
+	if len(newCompSet) == 0 {
+		return errors.New("cannot remove all components form an entity")
+	}
+	key := compKey{cType.ID(), id}
+	delete(m.compValues, key)
+	m.compValuesToDelete[key] = true
+	fromArchID, err := m.getArchIDForComponentsOrMakeIt(comps)
+	if err != nil {
+		return err
+	}
+	toArchID, err := m.getArchIDForComponentsOrMakeIt(newCompSet)
+	if err != nil {
+		return err
+	}
+	return m.moveEntityByArchetype(fromArchID, toArchID, id)
+}
+
+// GetComponentTypesForEntity returns all the component types that are currently on the given entity. Only types
+// are returned. To get the actual component data, use GetComponentForEntity.
+func (m *Manager) GetComponentTypesForEntity(id entity.ID) ([]component.IComponentType, error) {
+	archID, err := m.getArchetypeForEntity(id)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.GetComponentTypesForArchID(archID), nil
+}
+
+// GetComponentTypesForArchID returns the set of components that are associated with the given archetype id.
+func (m *Manager) GetComponentTypesForArchID(archID archetype.ID) []component.IComponentType {
+	return m.archIDToComps[archID]
+}
+
+// GetArchIDForComponents returns the archetype ID that has been assigned to this set of components.
+// If this set of components does not have an archetype ID assigned to it, an error is returned.
+func (m *Manager) GetArchIDForComponents(components []component.IComponentType) (archetype.ID, error) {
+	if len(components) == 0 {
+		return 0, errors.New("must provide at least 1 component")
+	}
+	if err := sortComponentSet(components); err != nil {
+		return 0, err
+	}
+	for archID, comps := range m.archIDToComps {
+		if isComponentSetMatch(comps, components) {
+			return archID, nil
+		}
+	}
+	return 0, errorArchIDNotFound
+}
+
+// GetEntitiesForArchID returns all the entities that currently belong to the given archetype ID.
+func (m *Manager) GetEntitiesForArchID(archID archetype.ID) []entity.ID {
+	active, err := m.getActiveEntities(archID)
+	if err != nil {
+		// TODO: This shouldn't be a panic. The error should be returned.
```

Comment:
I added a ticket. This is going to end up impacting the query code. Ultimately, I think the best course of action will be to return an error from this function, but since that will have wider implications that just the ECB, I want to leave it for another PR.